File: Programmers/Level_2/땅따먹기.py
'''
https://programmers.co.kr/learn/courses/30/lessons/12913

0행 부터 n행 까지 단방향으로의 최댓값만 찾으면 되는줄 알고 한참 헤맨 문제
역방향으로 가도 최댓값이 되도록 하는 값을 찾는게 문제의 숨은 의도였다

전 단계를 previous, 현재 단계를 current라고 한다면,
previous에서 밟았던 열을 current에서 또 밟으면 안 되므로 해당 열을 idx로 두고
previous의 최댓값의 인덱스가 idx와 동일하다면 2번째로 큰 수,
다르다면 가장 큰 수를 current[idx] 값과 더해서 저장한다
그러면 current까지 최대한 크게 오는 값 들을 알 수 있다
'''


def solution(land):

    def selectMax(row, idx):
        if idx == row.index(max(row)):
            return sorted(row)[-2]
        else:
            return sorted(row)[-1]

    for row in range(1, len(land)):
        for idx in range(4):
            land[row][idx] += selectMax(land[row-1], idx)

    return max(land[-1])

# print(solution([[1,2,3,5],[5,6,7,8],[4,3,2,1]]))

File: Programmers/Level_2/Swift_Code/Swift_Code/땅따먹기.swift
//
//  땅따먹기.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/03/17.
//

/*
 Python보다 Swift가 시간을 더 잡아먹는걸 다시 한 번 느꼈다
 시간 초과 1: O(NM)
 시간 초과 2: O(NM)
 최종 솔루션: O(N+M)
 
 전 단계에서 현재 열의 값을 제외한 값 중에서 가장 큰 값과 현재 행 현재 열의 값과 더해가면서
 마지막 행의 최댓값을 리턴
 */

import Foundation

// MARK: - 시간 초과 1
//func selectMax(_ previous: [Int], _ idx: Int) -> Int {
//    if previous.max() == previous[idx] {
//        return previous.sorted()[previous.endIndex - 2]
//    } else {
//        return previous.sorted()[previous.endIndex - 1]
//    }
//}

// MARK: - 시간 초과 2
//func solution(_ land: [[Int]]) -> Int {
//    var dp = [[Int]](repeating: [Int](repeating: 0, count: 4), count: land.count)
//    dp[0] = land[0]
//
//    for current in 1..<land.count {
//        for idx in 0...3 {
//            var previous = dp[current - 1]
//            previous.remove(at: idx)
//            dp[current][idx] = land[current][idx] + previous.max()!
//        }
//    }
//
//    return dp.last!.max()!
//}

func solution(_ land: [[Int]]) -> Int {
    var land = land
    
    for i in 1..<land.count {
        land[i][0] += max(land[i-1][1], land[i-1][2], land[i-1][3])
        land[i][1] += max(land[i-1][0], land[i-1][2], land[i-1][3])
        land[i][2] += max(land[i-1][0], land[i-1][1], land[i-1][3])
        land[i][3] += max(land[i-1][0], land[i-1][1], land[i-1][2])
    }

    return land.last!.max()!
}

// print(solution([[1,2,3,5],[5,6,7,8],[4,3,2,1]]))
